### PR TITLE
#358 - added support for synchronizing pruned initial `RecordsWrite`

### PR DESCRIPTION
**Pull request by thehenrytsai**
1. Changed LevelDB data store to separate data by tenant - prerequisite for adding the support
1. Added support for synchronizing pruned initial RecordsWrite through `synchronizePrunedInitialRecordsWrite()`
1. Refactored LevelDB data store

Note that we are unable to use the approach of relaxing `RecordsWriteHandler` directly because it would allow unauthorized access of data. See bug #359 for demonstrating the need to further lock down the `RecordsWriteHandler`.

Not super proud/happy with the approach taken, but it gets us over the finish line for now and it's not entirely messy. All ears for improvements or better approaches.

### CODE REVIEW

**Comment by csuwildcat** [comment]
Thank you, Henry!

**Review by mistermoe — DISMISSED** [review]
surgical! 🏅

File: src/interfaces/records/handlers/pruned-initial-records-write.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * Handler for initial `RecordsWrite` that already have its data pruned thus cannot supply data stream for it.
+ * NOTE: This is intended to be ONLY used by sync.
+ */
+export class PrunedInitialRecordsWriteHandler extends RecordsWriteHandler {
+  /**
+   * Stores the given message without storing the associated data.
+   * Requires `dataCid` to exist in the `RecordsWrite` message given.
+   */
+  public async storeMessage(
```
Comment (diehuxx) [inline src/interfaces/records/handlers/pruned-initial-records-write.ts:18]: I agree with you that this isn't my favorite approach. Inheritance always puts me edge. I don't want to block this PR, so I'll percolate on other ideas for it post-Miami.

**Review by diehuxx — DISMISSED** [review]

File: src/store/data-store-level.ts [hunk 2]
```diff
@@ -18,8 +15,8 @@ const PLACEHOLDER_VALUE = new Uint8Array();
  * Leverages LevelDB under the hood.
  *
  * It has the following structure (`+` represents a sublevel and `->` represents a key->value pair):
- *   'data' + <dataCid> -> <data>
- *   'host' + <dataCid> + <tenant> + <messageCid> -> PLACEHOLDER_VALUE
+ *   'data' + <tenant> + <dataCid> -> <data>
+ *   'host' + <tenant> + <dataCid> + <messageCid> -> PLACEHOLDER_VALUE
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/store/data-store-level.ts:19]: Thanks for the great pick up!

File: src/store/data-store-level.ts [hunk 3]
```diff
@@ -51,16 +48,12 @@ export class DataStoreLevel implements DataStore {
   }
 
   async put(tenant: string, messageCid: string, dataCid: string, dataStream: Readable): Promise<PutResult> {
-    const tenantsForData = await this.blockstore.partition(HOST_PARTITION);
-    const messagesForTenant = await tenantsForData.partition(dataCid);
-    const messages = await messagesForTenant.partition(tenant);
-
-    await messages.put(messageCid, PLACEHOLDER_VALUE);
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/store/data-store-level.ts:58]: Yeah, no change in purpose, I renamed it based on what I think it is doing after reading the code, 'coz I also found it not obvious.

File: src/interfaces/records/handlers/pruned-initial-records-write.ts (hunk 1 above)
Comment (thehenrytsai) [reply src/interfaces/records/handlers/pruned-initial-records-write.ts:18]: Looking forward to it!

File: src/dwn.ts [hunk 4]
```diff
@@ -135,6 +141,28 @@ export class Dwn {
...
+    const isTenant = await this.tenantGate.isTenant(tenant);
+    if (!isTenant) {
+      return new MessageReply({
+        status: { code: 401, detail: `${tenant} is not a tenant` }
+      });
+    }
+
+    try {
+      // consider to push this down to individual handlers
+      Message.validateJsonSchema(message);
```
[the remark this answers is not in the document]
Comment (thehenrytsai) [reply src/dwn.ts:169]: Done. Added a test alsol

File: tests/interfaces/records/handlers/records-delete.spec.ts [hunk 5]
```diff
@@ -132,14 +132,14 @@ describe('RecordsDeleteHandler.handle()', () => {
     });
 
     it('should only write the data once even if written by multiple messages', async () => {
```
Comment (diehuxx) [inline tests/interfaces/records/handlers/records-delete.spec.ts:134]: Not really relevant to the PR, but do you know why we have this test inside `handlers/records-delete.spec.ts`? Doesn't look like we use `RecordsDelete` or do deleting of any kind in this test.

**Pull request merged**